### PR TITLE
lint README and add gh workflow to do so

### PR DESCRIPTION
**Pull request by naseemkullah**
Signed-off-by: naseemkullah <naseem@transit.app>

#### Checklist

- [x] run `npm run test` and `npm run benchmark`
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message and code follows the [Developer's Certification of Origin](https://github.com/fastify/.github/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-11)
      and the [Code of conduct](https://github.com/fastify/.github/blob/master/CODE_OF_CONDUCT.md)

### CODE REVIEW

**Comment by mcollina** [comment]
Thanks! This is actually nice. 

```
docs/Encapsulation.md:22:1 MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
```

**Comment by naseemkullah** [comment]
> Thanks! This is actually nice.
> 
> ```
> docs/Encapsulation.md:22:1 MD004/ul-style Unordered list style [Expected: asterisk; Actual: dash]
> ```

My pleasure! Fixed, pardon the force push! :)

**Review by RafaelGSS — APPROVED** [review]
LGTM.

File: CONTRIBUTING.md
Comment (jsumners) [inline CONTRIBUTING.md:7]: Why has punctuation been removed from these headings?

File: PROJECT_CHARTER.md
Comment (jsumners) [inline PROJECT_CHARTER.md:64]: Why has this changed?

File: README.md
Comment (jsumners) [inline README.md:164]: Why the addition of `<` and `>`?

File: docs/Ecosystem.md
Comment (jsumners) [inline docs/Ecosystem.md:7]: Why wasn't the next subsection also changed to 3 instead of 4?

File: docs/Errors.md
Comment (jsumners) [inline docs/Errors.md:53]: 4 spaces is actually significant in Markdown. Is this change correct?

File: docs/Logging.md
Comment (jsumners) [inline docs/Logging.md:172]: Why the `<` and `>`?

File: docs/Middleware.md
Comment (jsumners) [inline docs/Middleware.md:32]: These should remain together.

File: docs/Plugins-Guide.md
Comment (jsumners) [inline docs/Plugins-Guide.md:364]: These should remain together.

File: docs/Plugins-Guide.md
Comment (jsumners) [inline docs/Plugins-Guide.md:387]: All headings and manual anchors should remain together. I'm going to stop leaving comments on them, but this is definitely not the last instance of this issue.

File: docs/Reply.md
Comment (jsumners) [inline docs/Reply.md:504]: More seemingly random `<` and `>` additions.

File: docs/Routes.md
Comment (jsumners) [inline docs/Routes.md:267]: This change definitely seems incorrect.

File: docs/Routes.md
Comment (jsumners) [inline docs/Routes.md:260]: The emphasis on "error"  in the last sentence has been lost.

File: docs/Server.md
Comment (jsumners) [inline docs/Server.md:35]: Why is this file different? The other got changed to `-` for lists, but this one to `*`?

File: docs/Style-Guide.md
Comment (jsumners) [inline docs/Style-Guide.md:7]: More lost punctuation.

File: docs/Testing.md
Comment (jsumners) [inline docs/Testing.md:180]: Lost punctuation.

File: docs/Testing.md
Comment (jsumners) [inline docs/Testing.md:231]: Lost punctuation.

File: docs/TypeScript.md
Comment (jsumners) [inline docs/TypeScript.md:450]: Is this level change correct?

File: docs/TypeScript.md
Comment (jsumners) [inline docs/TypeScript.md:1089]: Looks to me like there are mismatched back ticks here in the original document. Not sure why this change also removes a space between "be" and the last inlined code snippet.

**Review by jsumners — CHANGES_REQUESTED** [review]
I'm extremely hesitant to accept these automated changes. There are some significant formatting changes I have called out, and very likely many that I haven't.

**Comment by mcollina** [comment]
Then let's not land this.

**Review by mcollina — CHANGES_REQUESTED** [review]
I don't think these automated changes are good enough for us.

**Comment by naseemkullah** [comment]
In fact most thing @jsumners pointed it were manual errors on my end that the automation did not touch. I'll try again, this time only with automated changes and add any remaining broken rules to the ignored rules file.

File: docs/Logging.md
Comment (naseemkullah) [reply docs/Logging.md:172]: https://github.com/DavidAnson/markdownlint/blob/v0.22.0/doc/Rules.md#md034

We can ad md034 to the ignore list if you prefer? Please advise.

File: README.md
Comment (naseemkullah) [reply README.md:162]: https://github.com/DavidAnson/markdownlint/blob/v0.22.0/doc/Rules.md#md034

We can ad md034 to the ignore list if you prefer? Please advise.

File: CONTRIBUTING.md
Comment (naseemkullah) [reply CONTRIBUTING.md:7]: Due to https://github.com/DavidAnson/markdownlint/blob/v0.22.0/doc/Rules.md#md026---trailing-punctuation-in-heading

Would you like this rule ignored or that we reword the headings slight i,e, "About" and "Become a collaborator"?

File: README.md
Comment (naseemkullah) [reply README.md:162]: Added md034 to ignored rules.

File: docs/Reply.md
Comment (naseemkullah) [reply docs/Reply.md:458]: Its flagging bare URLs, maybe what we should do is:

```
- [discussion about this feature](https://github.com/fastify/fastify/issues/1864) 
- [definition of thenables](https://promisesaplus.com)
- [signature](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/then)
```

WDYT?

**Comment by naseemkullah** [comment]
More rules have been ignored and a much lighter touch has been applied, PTAL.

File: README.md
Comment (naseemkullah) [reply README.md:162]: We could also do this:

```
Benchmarks taken using [fastify benchmarks](https://github.com/fastify/benchmarks)
```

WDYT?

File: docs/Logging.md
Comment (naseemkullah) [reply docs/Logging.md:172]: We could also:

```
See [Pino documentation](https://getpino.io/#/docs/redaction) for more details.
```

WDYT?

File: .github/workflows/lint-md.yml
Comment (salmanm) [inline .github/workflows/lint-md.yml:3]: I'd add a `paths` filter

Comment (naseemkullah) [reply .github/workflows/lint-md.yml:3]: Good idea!

File: README.md
Comment (jsumners) [reply README.md:162]: > Rationale: Without angle brackets, the URL isn't converted into a link by many markdown parsers.

Except, they are on GitHub and the tooling we to build our hosted site. So I don't see the utility of the rule.

File: docs/Reply.md
Comment (jsumners) [reply docs/Reply.md:458]: I think the author chose to use bare links instead of titled links for a reason. I just don't think this rule is needed.

**Comment by jsumners** [comment]
Can we re-start this PR in a new one that addresses files in smaller, easier to digest, chunks? For example, maybe start off with the main README and one or two other documents. That would make it easier to review the rules and agree upon them before they are applied across everything. As we agree on the initial set of documents, a few more can be added, and so on until we have covered them all. I realize it will take more time, but this PR is really too much to review.

**Comment by naseemkullah** [comment]
Sure!

**Comment by naseemkullah** [comment]
PTAL @jsumners this PR has been restarted from scratch, only README is touched and most rules are turned off for the time being. Just blank lines and spaces. :)

File: .markdownlint.json
Comment (jsumners) [inline .markdownlint.json:18]: These are mysterious. Does this tool support a yaml configuration file so that comments can be added to describe each one?

File: README.md
Comment (jsumners) [inline README.md:78]: I don't see any problem with the changes to this file.

File: .markdownlint.json
Comment (naseemkullah) [reply .markdownlint.json:18]: I believe so! Stay tuned.

Comment (naseemkullah) [reply .markdownlint.json]: done ✅

File: .markdownlint.yaml
Comment (jsumners) [inline .markdownlint.yaml:1]: Thank you. This will help out a lot. Can you also add a description at the top of the file that notes where these rules come from and how to research them (link to docs)?

Comment (naseemkullah) [reply .markdownlint.yaml:1]: Sure!

Comment (naseemkullah) [reply .markdownlint.yaml:4]: ✅

**Review by jsumners — APPROVED** [review]
LGTM so far

**Review by Eomm — APPROVED** [review]

**Review by RafaelGSS — APPROVED** [review]
It's a better approach, for sure. LGTM.

**Comment by mcollina** [comment]
Would you mind to add a script in package.json to run the markdown linting?

**Comment by naseemkullah** [comment]
> Would you mind to add a script in package.json to run the markdown linting?

Good idea

**Comment by naseemkullah** [comment]
> Would you mind to add a script in package.json to run the markdown linting?

done!

File: package.json
Comment (jsumners) [inline package.json:24]: This tool is really annoying. I can't find any documentation about what options it accepts in the configuration dotfile to determine if the target file list can be specified there instead of repeated throughout our code base.

As an example of what I'm looking for, https://node-tap.org/docs/configuring/ clearly shows how to generate a default configuration file. It also shows that `--files` can be added to a configuration like so:

```yaml
files:
	- index.test.js
	- 'test/**/*.test.js'
```

Do you know if this same thing can be done with the Markdown lint tool?

Comment (naseemkullah) [reply package.json:24]: I'm not sure, I'm sorry you are annoyed, I can close this PR, the purpose was to improve document quality but the amount of grief it seems to cause you does not warrant it.

Comment (jsumners) [reply package.json:24]: I'm not against the PR. I'm just frustrated with the tool's lack of transparency. As best I can tell, `markdownlint-cli` is the thing that reads the configuration and kicks off the linter. But that tool's readme doesn't provide a list, or a link to a list, of all possible configuration options.

Comment (naseemkullah) [reply package.json:24]: That's true.

I've filed:
https://github.com/igorshubovych/markdownlint-cli/issues/158
https://github.com/igorshubovych/markdownlint-cli/issues/159

maybe see what they say and can decide if the tool will be worth fastify maintainers' while.

**Comment by DavidAnson** [comment]
If you want to be paranoid, it may be desirable to explicitly list the enabled rules in the configuration as well so that new rules aren't applied unexpectedly.

Additionally, you may choose to set `default:false` and then list ONLY the enabled rules so that it's exceedingly clear what IS applied (vs. how it looks now listing what is NOT being applied).

I have also commented on both issues opened against the CLI. Thanks!

File: package.json
Comment (naseemkullah) [reply package.json]: I've ported to `markdownlint-cli2` PTAL and let me know your preference!

The files we want to lint still need to be specified every time it is invoked though.

File: package.json
Comment (DavidAnson) [inline package.json:154]: Nit: Missing '^'

Comment (naseemkullah) [reply package.json]: thanks!

**Review by jsumners — COMMENTED** [review]
What is the difference with these new changes?

**Comment by naseemkullah** [comment]
> What is the difference with these new changes?

Not much in fact.

The new underlying CLI has more config options, none of which we use today but they are easy to find in the README.

Also thought I'd give it a try as the maintainer of both this CLI and it's underlying library (markdownlint) @DavidAnson has been very responsive to our feedback and may add a files config option in the future.

File: package.json
Comment (Eomm) [reply package.json]: The openjs foundation uses this tool [remark](https://www.npmjs.com/package/remark) that seems more documented to lint MD files

**Review by jsumners — APPROVED** [review]

**Pull request closed without merging**
